### PR TITLE
Pod eviction grace period flag

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Add the last remaining flag from kubelet eviction proposal.

/cc @vishh

### CODE REVIEW

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -257,4 +257,5 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EvictionHard, "eviction-hard", s.EvictionHard, "A set of eviction thresholds (e.g. memory.available<1Gi) that if met would trigger a pod eviction.")
 	fs.StringVar(&s.EvictionSoft, "eviction-soft", s.EvictionSoft, "A set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a pod eviction.")
 	fs.StringVar(&s.EvictionSoftGracePeriod, "eviction-soft-grace-period", s.EvictionSoftGracePeriod, "A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.")
+	fs.Int32Var(&s.EvictionMaxPodGracePeriod, "eviction-max-pod-grace-period", s.EvictionMaxPodGracePeriod, "Maximum allowed grace period (in seconds) to use when terminating pods in response to a soft eviction threshold being met.")
```
Comment (vishh) [inline cmd/kubelet/app/options/options.go]: Why not take in a `Duration`?

File: docs/admin/kubelet.md [hunk 2]
```diff
@@ -88,6 +88,7 @@ kubelet
       --event-burst=10: Maximum size of a bursty event records, temporarily allows event records to burst to this number, while still not exceeding event-qps. Only used if --event-qps > 0
       --event-qps=5: If > 0, limit event creations per second to this value. If 0, unlimited.
       --eviction-hard="": A set of eviction thresholds (e.g. memory.available<1Gi) that if met would trigger a pod eviction.
+      --eviction-max-pod-grace-period=0: Maximum allowed grace period (in seconds) to use when terminating pods in response to a soft eviction threshold being met.
```
Comment (vishh) [inline docs/admin/kubelet.md]: This is set to `0` by default. Is that desirable. Should we choose a better default? How about `30s`?

**Comment by vishh** [comment]
Just a couple of comments, otherwise LGTM

File: docs/admin/kubelet.md (hunk 2 above)
Comment (derekwaynecarr) [inline docs/admin/kubelet.md]: You asked for 0 as the default?  In that manner a soft eviction threshold being met for a specified duration had the same eviction behavior as a hard eviction threshold.  Maybe we need a -1 to denote just use what the pod evicted specified.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (derekwaynecarr) [inline cmd/kubelet/app/options/options.go]: We can only support timeouts in second precision.  We can take a duration with that caveat if you think its a must.

File: docs/admin/kubelet.md (hunk 2 above)
Comment (vishh) [inline docs/admin/kubelet.md]: Yeah `-1` makes sense.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (vishh) [inline cmd/kubelet/app/options/options.go]: it just makes it easy to express minutes in addition to seconds. FWIW, we have other flags that take duration as well and are most likely using a second granularity underneath.

Comment (smarterclayton) [inline cmd/kubelet/app/options/options.go]: We already have grace-period today in the CLI, and it is always seconds.  We have stated that things should say "-seconds" in their API.  I don't think we make the argument that everything should be duration, just things that that have wildly disparate values.

If this was a duration, we'd have to chop and round it.  I'd be ok with that, as long as it translated to seconds throughout the code.

Comment (derekwaynecarr) [inline cmd/kubelet/app/options/options.go]: So this is consistent with --grace-period being in seconds, so keeping as is.

File: docs/admin/kubelet.md (hunk 2 above)
Comment (derekwaynecarr) [inline docs/admin/kubelet.md]: Updated description to describe a negative value is defer to pod spec.

**Comment by derekwaynecarr** [comment]
hopefully, this is good to go.

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/24620

**Comment by derekwaynecarr** [comment]
Fixed deep copy generation.

**Pull request merged**